pkg/algorithm/sku_scheduler.go

```diff
@@ -0,0 +1,512 @@
+			break
+		}
+	}
+
+	for {
+		for freeCellIndex < int32(len(freeCells)) {
+			freeCell = freeCells[freeCellIndex]
+			currentCellIndices[searchCellIndex] = freeCellIndex
+			if searchCellIndex == 0 {
+				currentAffinity[searchCellIndex] = freeCell
+			} else {
+				currentAffinity[searchCellIndex] = findLCA(freeCell, currentAffinity[searchCellIndex-1])
+				// pruning: if the current LCA has been higher than the lowest ever,
+				// the node will be skipped
+				if (currentAffinity[searchCellIndex] == nil && bestAffinity < highestLevel) ||
+					(currentAffinity[searchCellIndex] != nil && currentAffinity[searchCellIndex].GetLevel() > bestAffinity) {
+					freeCellIndex++
+					continue
+				}
+			}
+			if searchCellIndex == pod.CellsPerPod.CellNumber-1 {
+				foundOptimalAffinity := false
+				bestAffinity, foundOptimalAffinity = checkCurrentCells(
+					currentAffinity[len(currentAffinity)-1].GetLevel(),
+					freeCells,
+					currentCellIndices,
+					bestAffinity,
+					bestAffinityCells,
+					bestAffinityCellIndices,
+					optimalAffinity)
+				if foundOptimalAffinity {
+					// early stop: return if the solution is optimal (i.e., all buddies)
+					return bestAffinityCells
+				}
+			} else {
+				searchCellIndex++
+			}
+			freeCellIndex++
+		}
+		searchCellIndex--
+		if searchCellIndex < 0 {
+			if bestAffinity == highestLevel {
+				// Unreachable
+				panic(fmt.Sprintf("Assert Failure: failed to allocate %v cells in cell %v", pod.CellsPerPod.CellNumber, within.address))
+			}
+			return bestAffinityCells
+		}
+		freeCellIndex = currentCellIndices[searchCellIndex] + 1
+	}
+}
+
+func (s *skuScheduler) getNodeLevel() CellLevel {
+	for l := CellLevel(1); l <= CellLevel(len(s.chainCellList)); l++ {
+		if s.chainCellList[l][0].AtOrHigherThanNode() {
+			return l
+		}
+	}
+	return -1
+}
+
+// getFreeCellsAtLevel collects free cells and preemptible cells at given level according to the priority.
+func getFreeCellsAtLevel(
+	cell Cell,
+	level CellLevel,
+	priority CellPriority,
+	allocatedCells CellList,
+	freeCells CellList,
+	preemptibleCells CellList) (
+	CellList, CellList) {
+
+	if cell.GetLevel() > level {
+		for _, c := range cell.GetChildren() {
+			freeCells, preemptibleCells = getFreeCellsAtLevel(
+				c, level, priority, allocatedCells, freeCells, preemptibleCells)
+		}
+	} else if cell.GetLevel() == level {
+		isAllocated := false
+		for _, c := range allocatedCells {
+			if isAncestor(cell, c) || isAncestor(c, cell) {
+				isAllocated = true
+				break
+			}
+		}
+		if !isAllocated {
+			if cell.GetPriority() == freePriority {
+				freeCells = append(freeCells, cell)
+			} else if cell.GetPriority() < priority {
+				preemptibleCells = append(preemptibleCells, cell)
+			}
+		}
+	}
+	return freeCells, preemptibleCells
+}
+
+// checkCurrentCells checks if the currently picked cells have the lowest LCA.
+// It also checks if the solution is optimal (if the leaf cells are all buddies).
+func checkCurrentCells(
```

Comment:
The function name of checkCurrentCells is vague. Suggest to use names like "checkAffinityLowerAndOptimal".

Reply:
renamed to `checkOptimalAffinityForCells`